### PR TITLE
WIP - Token base authentication

### PR DESCRIPTION
**Pull request by aleklisi**

### CODE REVIEW

File: lib/mix/tasks/certs_dev.ex [hunk 1]
```diff
@@ -29,6 +29,7 @@ defmodule Mix.Tasks.Certs.Dev do
 
   @spec run(term) :: :ok
   def run(_) do
+    generate_p8_private_key()
```
Comment (rslota) [inline lib/mix/tasks/certs_dev.ex]: All functions here have `maybe_` keyword and characteristic (only generate new key if not present). I would suggest to stick with it. 

Also functions for certs specify type of the cert file, so this one also could do that. My proposition: `maybe_gen_apns_token`. This way we know what's that for.

File: lib/mix/tasks/certs_dev.ex [hunk 2]
```diff
@@ -146,4 +147,19 @@ defmodule Mix.Tasks.Certs.Dev do
...
+           "ecparam",
+           "-name",
+           "secp256r1",
+           "-genkey",
+           "-noout",
+           "-out",
+           "priv/apns/token.p8"
+         ]) do
+      {"", 0} -> :ok
+      _ -> raise "Cannot generate p8 private key!!!"
```
Comment (rslota) [inline lib/mix/tasks/certs_dev.ex]: I would strongly recommend passing the error code / reason.

File: lib/sparrow/apns/token_bearer.ex [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  def get_token() do
+    :ets.lookup_element(@tab_name, :apns_token, 2)
+  end
+
+  @spec init(token_bearer_opts) :: {:ok, Sparrow.APNS.TokenBearer.State.t()}
+  def init(opts) do
+    key_id = Keyword.get(opts, :key_id)
+    team_id = Keyword.get(opts, :team_id)
+    p8_file_path = Keyword.get(opts, :p8_file_path)
+    refresh_token_time_in_minutes = Keyword.get(opts, :refresh_token_time_in_minutes, 50)
```
Comment (rslota) [inline lib/sparrow/apns/token_bearer.ex]: The default timeout value should be defined as constant.

File: lib/sparrow/apns/token_bearer.ex [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    state =
+      Sparrow.APNS.TokenBearer.State.new(
+        key_id,
+        team_id,
+        p8_file_path,
+        refresh_token_time_in_minutes
+      )
+
+    @tab_name = :ets.new(@tab_name, [:set, :protected, :named_table])
+    _ = handle_info(:update_token, state)
```
Comment (rslota) [inline lib/sparrow/apns/token_bearer.ex]: Calling gen_server's callbacks manually is a bad practice and should never be done, unless maybe it's the last function call (we want to use the actual return value as local return). There should be function `update_token` that could be called in both places you need it.

File: lib/sparrow/apns/token_bearer.ex [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+          inspect(ets_del)
+        }"
+      end)
+  end
+
+  @spec handle_info(:update_token | any, Sparrow.APNS.TokenBearer.State.t()) ::
+          {:noreply, Sparrow.APNS.TokenBearer.State.t()}
+  def handle_info(:update_token, state) do
+    state.refresh_token_time_in_minutes
+    |> minutes_to_miliseconds
```
Comment (rslota) [inline lib/sparrow/apns/token_bearer.ex]: Just use `:timer.minutes/1` :)

File: lib/sparrow/apns/token_bearer.ex [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    {:noreply, state}
+  end
+
+  @spec set_new_token(Sparrow.APNS.TokenBearer.State.t()) :: true
+  defp set_new_token(state) do
+    {:ok, token, _} = new_JWT(state.key_id, state.team_id, state.p8_file_path)
+    :ets.insert(@tab_name, {:apns_token, token})
+  end
+
+  @spec new_JWT(String.t(), String.t(), String.t()) ::
```
Comment (rslota) [inline lib/sparrow/apns/token_bearer.ex]: This name doesn't conform to official coding standards - my suggestion: `new_jwt_token`.

File: lib/sparrow/apns/token_bearer.ex [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+          {:error, reason :: any} | {:ok, bearer_token, claims}
+  defp new_JWT(key_id, team_id, p8_file_path) do
+    signer = new_signer("ES256", key_id, p8_file_path)
```
Comment (rslota) [inline lib/sparrow/apns/token_bearer.ex]: It would be nice to define `@apns_jwt_alg = "ES256"`

File: test/apns/token_bearer_test.exs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      refresh_token_time_in_minutes: @refresh_token_time_in_minutes
+    ]
+
+    {:ok, pid} = GenServer.start_link(Sparrow.APNS.TokenBearer, opts)
+    {:ok, token_bearer_pid: pid}
+  end
+
+  test "token gets updated", context do
+    token_before_update = Sparrow.APNS.TokenBearer.get_token()
+    send(context[:token_bearer_pid], :update_token)
```
Comment (rslota) [inline test/apns/token_bearer_test.exs]: I would suggest changing the refresh interval to milliseconds, this way we could test it properly, by setting to e.g. 50ms and then sleeping 100ms. To be fair, in our API it doesn't matter really whether we provide minutes or milliseconds (the application itself may do this translation).

File: lib/sparrow/apns/token_bearer.ex (hunk 7 above)
Comment (aleklisi) [reply lib/sparrow/apns/token_bearer.ex]: you mean?
```elixir
@apns_jwt_alg "ES256"
```

File: lib/sparrow/h2_worker.ex [hunk 9]
```diff
@@ -59,8 +83,14 @@ defmodule Sparrow.H2Worker do
 
   @spec terminate(reason, state) :: :ok
   def terminate(reason, state) do
-    H2Adapter.close(state.connection_ref)
-    _ = Logger.info(fn -> "action=terminate, reason=#{inspect(reason)}" end)
+    case state.connection_ref do
```
Comment (rslota) [inline lib/sparrow/h2_worker.ex]: This would look better if we matched this in function head.

**Pull request closed without merging**